File: frontend/smartai_v2/api/analytics.py
"""Analytics API client (NL filter/summary/chart + per-question breakdown)."""
from __future__ import annotations

from .client import post_json, get_json, delete_json


async def nl_query(task_id: str, question: str, mode: str,
                   *, token: str | None = None) -> dict:
    """POST /analytics/{task_id}/query → {mode, ...mode-specific fields}"""
    return await post_json(
        f"/analytics/{task_id}/query",
        {"question": question, "mode": mode},
        token=token,
    )


async def per_question(task_id: str, q_id: str,
                       *, token: str | None = None) -> dict:
    return await get_json(f"/analytics/{task_id}/per_question/{q_id}", token=token)


async def reset_per_question_cache(task_id: str, q_id: str,
                                   *, token: str | None = None) -> dict:
    return await delete_json(f"/analytics/{task_id}/per_question/{q_id}/cache", token=token)

File: frontend/app/src/api/analytics.ts
import { EMPTY_FILTER_INTENT } from "@/lib/taskFilterIntent";
import { isGroundedExecution } from "@/lib/groundedAsk";
import type { GroundedAskExecution } from "@/types";
import { deleteJSON, getJSON, postJSON } from "./client";
import type { AnalyticsMode, AnalyticsResult, FilterIntentResult, FilterIntentSurface, PerQuestionBreakdown } from "@/types";

// Model-backed operations can outlast the ordinary 30-second data-request limit.
export const ANALYTICS_REQUEST_TIMEOUT_MS = 300_000;

export function runAnalyticsQuery(
  taskId: string,
  question: string,
  mode: AnalyticsMode,
  history?: string[],
): Promise<AnalyticsResult> {
  return postJSON<AnalyticsResult>(`/analytics/${taskId}/query`, { question, mode, ...(history?.length ? { history: history.slice(-4) } : {}) }, { timeout: ANALYTICS_REQUEST_TIMEOUT_MS });
}

export interface AskContext { studentId?: string; history?: string[] }
export async function runGroundedAsk(taskId: string | undefined, question: string, surface: string,
  signal?: AbortSignal, context?: AskContext): Promise<GroundedAskExecution> {
  const endpoint = taskId ? `/analytics/${encodeURIComponent(taskId)}/ask` : "/analytics/ask";
  const result = await postJSON<GroundedAskExecution>(endpoint, { question, surface,
    ...(context?.studentId ? { context_student_id: context.studentId } : {}),
    ...(context?.history?.length ? { history: context.history.slice(-4) } : {}),
  }, { timeout: ANALYTICS_REQUEST_TIMEOUT_MS, ...(signal ? { signal } : {}) });
  if (!isGroundedExecution(result)) throw new Error("Invalid Ask query result");
  return result;
}
export async function interpretFilterIntent(taskId: string, question: string,
  surface: FilterIntentSurface, signal?: AbortSignal, context?: AskContext): Promise<FilterIntentResult> {
  const execution = await runGroundedAsk(taskId, question, surface, signal, context);
  return { ...EMPTY_FILTER_INTENT, recognized: execution.recognized, explanation: execution.explanation, execution };
}

export function getPerQuestionBreakdown(taskId: string, qId: string): Promise<PerQuestionBreakdown> {
  return getJSON<PerQuestionBreakdown>(`/analytics/${taskId}/per_question/${qId}`, { timeout: ANALYTICS_REQUEST_TIMEOUT_MS });
}

export function resetPerQuestionCache(taskId: string, qId: string): Promise<{ status: "cleared" }> {
  return deleteJSON<{ status: "cleared" }>(`/analytics/${taskId}/per_question/${qId}/cache`);
}
